### PR TITLE
Avoid String.format usages in hot paths

### PR DESCRIPTION
**Pull request by pettyjamesm**
## Description
Replaces usages of `String.format` in expected codepaths to improve performance. In particular `time` / `timestamp` related JSON serializers and casts to `varchar` improve dramatically with `timestamp(3)` casting throughput improving 2x, `time(3)` improving more than 30x, and `time with time zone(3)` improving ~40x (benchmark results: [jmh.morethan.io](https://jmh.morethan.io/?sources=https://gist.githubusercontent.com/pettyjamesm/706f34955e580d054606741ad41d3e3c/raw/2a905702d120ddb0a5b4ba58f28cfc60721b60fc/baseline-full.json,https://gist.githubusercontent.com/pettyjamesm/706f34955e580d054606741ad41d3e3c/raw/2a905702d120ddb0a5b4ba58f28cfc60721b60fc/improved-full.json)).

## Release notes

(x) This is not user-visible or docs only and no release notes are required.
( ) Release notes are required, please propose a release note for me.
( ) Release notes are required, with the following suggested text:

### CODE REVIEW

File: core/trino-spi/src/main/java/io/trino/spi/type/TimeZoneKey.java [hunk 1]
```diff
@@ -187,7 +188,7 @@ public static TimeZoneKey getTimeZoneKeyForOffset(long offsetMinutes)
         if (!(offsetMinutes >= OFFSET_TIME_ZONE_MIN && offsetMinutes <= OFFSET_TIME_ZONE_MAX)) {
             throw new TrinoException(INVALID_FUNCTION_ARGUMENT, "Invalid offset minutes " + offsetMinutes);
         }
-        TimeZoneKey timeZoneKey = OFFSET_TIME_ZONE_KEYS[((int) offsetMinutes) - OFFSET_TIME_ZONE_MIN];
+        TimeZoneKey timeZoneKey = OFFSET_TIME_ZONE_KEYS[(int) (offsetMinutes - OFFSET_TIME_ZONE_MIN)];
```
Comment (martint) [inline core/trino-spi/src/main/java/io/trino/spi/type/TimeZoneKey.java:191]: Unrelated change?

File: core/trino-main/src/test/java/io/trino/operator/scalar/timestamp/BenchmarkCastTimestampToVarchar.java [hunk 2]
```diff
@@ -0,0 +1,191 @@
...
+import static org.openjdk.jmh.annotations.Mode.Throughput;
+import static org.openjdk.jmh.annotations.Scope.Thread;
+
+@State(Thread)
+@OutputTimeUnit(SECONDS)
+@BenchmarkMode(Throughput)
+@Fork(1)
+@Warmup(iterations = 5, time = 1, timeUnit = SECONDS)
+@Measurement(iterations = 5, time = 1, timeUnit = SECONDS)
+public class BenchmarkCastTimestampToVarchar
```
Comment (martint) [inline core/trino-main/src/test/java/io/trino/operator/scalar/timestamp/BenchmarkCastTimestampToVarchar.java:69]: Why not benchmark the cast methods directly instead of going through the expression compiler and processor machinery?

File: core/trino-spi/src/main/java/io/trino/spi/type/SqlTimeWithTimeZone.java [hunk 3]
```diff
@@ -103,20 +103,38 @@ public int hashCode()
...
-            builder.append(".");
-            builder.append(format("%0" + precision + "d", scaledFraction));
+            builder.append('.');
+            builder.setLength(builder.length() + precision);
+
+            for (int index = builder.length() - 1; index > 8; index--) {
+                long temp = scaledFraction / 10;
+                int digit = (int) (scaledFraction - (temp * 10));
+                scaledFraction = temp;
+                builder.setCharAt(index, (char) ('0' + digit));
```
Comment (martint) [inline core/trino-spi/src/main/java/io/trino/spi/type/SqlTimeWithTimeZone.java:122]: Didn't we add similar code somewhere else recently? We may want to consider putting it somewhere we can reuse it, since it's not particularly trivial.

Comment (pettyjamesm) [reply core/trino-spi/src/main/java/io/trino/spi/type/SqlTimeWithTimeZone.java:122]: We did indeed. That change was to SqlTimestamp in the SPI which improved performance just for the JSON serialization path. I think there are a few problems that I think has caused this code to proliferate:
- SqlTimestamp and SqlTimestampWithTimeZone are SPI objects that seem to only exist in order to serialize values as JSON. They have similar, but slightly different logic from one another (the time zone part)
- Casting to timestamp (and with time zone) types are currently implemented in trino-main and also formatted to strings (even though casting to varchar needs a UTF8 slice) because the formatting logic is being shared for long and short timestamp types both with and without time zones and for conversions and in parts of the JSON type handling paths. So it’s a 2x2 grid of timestamp types each targeting at least 2 output representations, and not accessible from the SPI - meanwhile the SPI can’t host the implementation because it depends on the packing / unpacking implementations in main.
- Time with and without time zone is similar in Trino main, except that was already producing a slice directly (presumably for performance to avoid the intermediate string and since the format is simple enough to implement by hand) which means it also can’t live in the SPI because of the dependency on slice.
